### PR TITLE
Add CPU horizontal fusion pass for inception.

### PR DESCRIPTION
**Pull request by ayzhuang**

### CODE REVIEW

File: src/ngraph/runtime/cpu/cpu_external_function.cpp [hunk 1]
```diff
@@ -1141,6 +1143,7 @@ void runtime::cpu::CPU_ExternalFunction::build()
     pass_manager.register_pass<ngraph::pass::CommonSubexpressionElimination>();
     pass_manager.register_pass<ngraph::pass::CoreFusion>();
     pass_manager.register_pass<runtime::cpu::pass::CPUFusion>();
+    pass_manager.register_pass<runtime::cpu::pass::CPUInceptionHorizontalFusion>();
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/cpu_external_function.cpp]: do we need a separate pass? I would prefer if this is added to cpu_fusion. Also we dont need to say Inception here. This is more generic than the inception model.

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp [hunk 2]
```diff
@@ -1272,3 +1275,105 @@ void ngraph::runtime::cpu::pass::CPUFusion::construct_bounded_relu()
...
+        auto conv_bias_root = std::dynamic_pointer_cast<op::ConvolutionBias>(m.get_match_root());
+
+        //check if the node has been replaced
+        if (conv_bias_root->get_users().empty())
+        {
+            return false;
+        }
+
+        //check if 1x1 filter
+        auto m_filters_shape = conv_bias_root->get_input_shape(1);
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: I thought we could always do this optimization regardless of a filter size as long as all convolutions are using the same data?

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp [hunk 3]
```diff
@@ -1272,3 +1275,105 @@ void ngraph::runtime::cpu::pass::CPUFusion::construct_bounded_relu()
...
+            return false;
+        }
+
+        // get weights and bias from each CBR 1x1 and create Concat nodes
+        std::vector<std::shared_ptr<Node>> weights_nodes;
+        std::vector<std::shared_ptr<Node>> bias_nodes;
+        std::vector<std::shared_ptr<Node>> conv_bias_nodes;
+
+        for (auto u : m.get_pattern_map()[data_conv]->get_users())
+        {
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: can we add a check that input data is the same across all convolutions?

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp [hunk 4]
```diff
@@ -1272,3 +1275,105 @@ void ngraph::runtime::cpu::pass::CPUFusion::construct_bounded_relu()
...
+            NGRAPH_DEBUG << "not 1x1 filter\n";
+            return false;
+        }
+
+        // get weights and bias from each CBR 1x1 and create Concat nodes
+        std::vector<std::shared_ptr<Node>> weights_nodes;
+        std::vector<std::shared_ptr<Node>> bias_nodes;
+        std::vector<std::shared_ptr<Node>> conv_bias_nodes;
+
+        for (auto u : m.get_pattern_map()[data_conv]->get_users())
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: do we still want to do this opt if convolutions feed into a maxpool? we won't be able to optimize slices (i.e. re-use concat storage?) @jbobba @ayzhuang ?

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp [hunk 5]
```diff
@@ -1272,3 +1275,105 @@ void ngraph::runtime::cpu::pass::CPUFusion::construct_bounded_relu()
     auto m = std::make_shared<pattern::Matcher>(min, callback);
     this->add_matcher(m);
 }
+
+void ngraph::runtime::cpu::pass::CPUFusion::construct_horizontal_fusion()
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: @jbobba could you please suggest a more descriptive name? this is your forté 😄

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp [hunk 6]
```diff
@@ -1272,3 +1275,105 @@ void ngraph::runtime::cpu::pass::CPUFusion::construct_bounded_relu()
...
+        return std::dynamic_pointer_cast<ngraph::op::Concat>(n) ||
+               std::dynamic_pointer_cast<ngraph::op::MaxPool>(n);
+    };
+
+    auto data_conv = std::make_shared<pattern::op::Label>(
+        element::f32, Shape{1, 256, 35, 35}, is_concat_or_maxpool);
+    auto filters = std::make_shared<pattern::op::Label>(element::f32, Shape{64, 256, 1, 1});
+    auto bias = std::make_shared<pattern::op::Label>(element::f32, Shape{64});
+
+    auto conv_bias = std::make_shared<ngraph::op::ConvolutionBias>(data_conv,
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: Can we have ConvolutionBias + Relu or any other combo we can easily handle?

File: src/ngraph/runtime/cpu/pass/cpu_fusion.cpp (hunk 5 above)
Comment (jbobba) [reply src/ngraph/runtime/cpu/pass/cpu_fusion.cpp]: i'm ok with horizontal_fusion, may be construct_conv_horizontal_fusion

File: src/ngraph/runtime/cpu/pass/cpu_fusion.hpp [hunk 7]
```diff
@@ -71,6 +71,12 @@ class ngraph::runtime::cpu::pass::CPUFusion : public ngraph::pass::GraphRewrite
             construct_conv_bias_add();
             construct_conv_bias_add_relu();
             construct_bounded_relu();
+        }
```
Comment (jbobba) [inline src/ngraph/runtime/cpu/pass/cpu_fusion.hpp]: Incorrect merge?

File: src/ngraph/runtime/cpu/pass/cpu_horizontal_fusion.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        auto conv_bias_root = std::dynamic_pointer_cast<op::ConvolutionBias>(m.get_match_root());
+
+        //check if the node has been replaced
+        if (conv_bias_root->get_users().empty())
+        {
+            return false;
+        }
+
+        auto m_filters_shape = conv_bias_root->get_input_shape(1);
+        auto f1 = m_filters_shape[2];
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_horizontal_fusion.cpp]: h,w?

File: src/ngraph/runtime/cpu/pass/cpu_horizontal_fusion.cpp [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            {
+                continue;
+            }
+            weights_nodes.push_back(u->get_argument(1));
+            bias_nodes.push_back(u->get_argument(2));
+            conv_bias_nodes.push_back(u);
+        }
+        if (conv_bias_nodes.size() <= 1)
+        {
+            return false;
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_horizontal_fusion.cpp]: I wonder if we could add `NGRAPH_DEBUG` as to why we bailed out?

File: src/ngraph/runtime/cpu/pass/cpu_horizontal_fusion.cpp [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+            if (u->get_argument(0) != m.get_pattern_map()[data_conv])
+            {
+                NGRAPH_DEBUG << "conv_horizontal_fusion: data_conv is not input 0 for "
+                             << u->get_name() << "\n";
+                continue;
+            }
+            auto u_filters_shape = u->get_input_shape(1);
+            if (u_filters_shape[2] != f1 || u_filters_shape[3] != f2)
+            {
+                continue;
```
Comment (Krovatkin) [inline src/ngraph/runtime/cpu/pass/cpu_horizontal_fusion.cpp]: NGRAPH_DEBUG?

**Pull request merged**